app/code/community/Bolt/Boltpay/OrderCreationException.php

```diff
@@ -0,0 +1,214 @@
+     */
+    const E_BOLT_CART_HAS_EXPIRED_TMPL_SHIPPING_TAX       = '{"reason": "Shipping tax amount has changed", "old_value": "%d", "new_value": "%d"}';
+
+    /**
+     * Item price changes
+     */
+    const E_BOLT_ITEM_PRICE_HAS_BEEN_UPDATED         = 2001004;
+    const E_BOLT_ITEM_PRICE_HAS_BEEN_UPDATED_TMPL    = '{"product_id": "%d", "old_price": "%d", "new_price": "%d"}';
+
+    /**
+     * Items have become out of stock without back-orders enabled
+     */
+    const E_BOLT_OUT_OF_INVENTORY         = 2001005;
+    const E_BOLT_OUT_OF_INVENTORY_TMPL    = '{"product_id": "%d", "available_quantity": %d, "needed_quantity": %d}';
+
+    /**
+     * Error applying discount to the cart
+     */
+    const E_BOLT_DISCOUNT_CANNOT_APPLY                 = 2001006;
+    const E_BOLT_DISCOUNT_CANNOT_APPLY_TMPL_EXPIRED    = '{"reason": "This coupon has expired", "discount_code": "%s"}';
+    const E_BOLT_DISCOUNT_CANNOT_APPLY_TMPL_GENERIC    = '{"reason": "%s", "discount_code": "%s"}';
+
+    /**
+     * Invalid discount code used
+     */
+    const E_BOLT_DISCOUNT_DOES_NOT_EXIST        = 2001007;
+    const E_BOLT_DISCOUNT_DOES_NOT_EXIST_TMPL   = '{"discount_code": "%s"}';
+
+    /**
+     * Change in shipping price
+     */
+    const E_BOLT_SHIPPING_PRICE_HAS_BEEN_UPDATED           = 2001008;
+    const E_BOLT_SHIPPING_PRICE_HAS_BEEN_UPDATED_TMPL      = '{"reason": "Shipping total has changed", "old_value": "%d", "new_value": "%d"}';
+
+    /**
+     * @var int http response code that is to be returned
+     */
+    protected $httpCode;
+
+    /**
+     * @var string The json to be returned to Bolt associated with this exception
+     */
+    protected $json;
+
+    /**
+     * @var array All possible error codes supported by Bolt
+     */
+    protected static $validCodes = array(
+        self::E_BOLT_GENERAL_ERROR,
+        self::E_BOLT_ORDER_ALREADY_EXISTS,
+        self::E_BOLT_CART_HAS_EXPIRED,
+        self::E_BOLT_ITEM_PRICE_HAS_BEEN_UPDATED,
+        self::E_BOLT_OUT_OF_INVENTORY,
+        self::E_BOLT_DISCOUNT_CANNOT_APPLY,
+        self::E_BOLT_DISCOUNT_DOES_NOT_EXIST,
+        self::E_BOLT_SHIPPING_PRICE_HAS_BEEN_UPDATED
+    );
+
+    /**
+     * Bolt_Boltpay_InvalidTransitionException constructor.
+     *
+     * @param int    $code          The Bolt defined error code
+     * @param string $dataTemplate  specific Bolt error sub-category
+     * @param array  $dataValues    An array of values to be added to the data template
+     * @param string $message       The exception message to throw.
+     * @param Throwable|null $previous  [optional] The previously throwable used for exception chaining.
+     */
+    public function __construct($code = self::E_BOLT_GENERAL_ERROR, $dataTemplate = self::E_BOLT_GENERAL_ERROR_TMPL_GENERIC, array $dataValues = array(), $message = null, Throwable $previous = null)
+    {
+        // If code is invalid, we will use the generic message
+        if (!in_array($code, self::$validCodes)) {
```

Comment:
Hm, this seems error prone. i can imagine someone adding constant to this class without realizing you have to update `validCodes` array also. Do we need this logic?

Reply:
It is not error prone.  If the person forgets to add a valid code, they will get a generic error message with the reason for the error clearly stated in a generic way.  This will be very clear on any basic inspection of Datadog or Bugsnag

However, the person should not forget to register their new error code.  It is clearly commented and should be very obvious to do so to junior and mid-level developers.

Reply:
what do you think about removing this logic and `$validCodes`?